src/nxt_h1proto.c

```diff
@@ -946,8 +934,20 @@ nxt_h1p_request_body_read(nxt_task_t *task, nxt_http_request_t *r)
                   &tmp_name, b->file->fd);
 
         unlink((char *) tmp_name.start);
+
+    } else {
+        tmp_path = NULL;
+        tmp_name.length = 0;
```

Comment:
Hmm, there was comment that went along with this

```c
-        /* This initialization required for CentOS 6, gcc 4.4.7. */
-        tmp_path = NULL;
-        tmp_name.length = 0;
```

Is this still required?

The oldest version of GCC we _may_ currently support is 4.7.x for C11/GNU11

[4.6.x](https://gcc.gnu.org/onlinedocs/gcc-4.6.4/gcc/Standards.html#Standards) did have some preliminary support, but

> A fourth version of the C standard, known as C1X, is under development; GCC has limited preliminary support for parts of this standard, enabled with -std=c1x.

not via -std=c11/gnu11

[4.7.x](https://gcc.gnu.org/onlinedocs/gcc-4.7.4/gcc/Standards.html#Standards) was the first to introduce -std=c11/gnu11

> A fourth version of the C standard, known as C11, was published in 2011 as ISO/IEC 9899:2011. GCC has limited incomplete support for parts of this standard, enabled with -std=c11 or -std=iso9899:2011. (While in development, drafts of this standard version were referred to as C1X.)

Having said all that, I think the oldest version of GCC we actively use/used is 4.8.5 from CentOS 7 which is essentially EOL now...

Reply:
> Is this still required?

I removed it. I feel it's needed in the past because of the logic like this:
```
if (a) {
    b = alloc1();
} else {
    /* This initialization required for CentOS 6, gcc 4.4.7. */
    b = alloc2()
}

if (b == NULL) {
    return NULL;
}

if (a) {
    use b;
} else {
    use b;
}
```

I merged the two condition branches, something like this:
```
if (a) {
    b = alloc1();
    if (b == NULL) {
        return NULL;
    }

    /* more logic will be added in the subsequent patch.
    use b;

} else {
    b = alloc2();
    if (b == NULL) {
        return NULL;
    }
}
```

But unfortunately, I can't reproduce the warning in the current code.
I feel it's impossible to cause such a warning with the new change.

Reply:
For interest...

Current GCC, 14.x, at least doesn't complain if you remove those dummy initialisations, so it got smarter...

Also just for clarity, the problem was

```c
if (body_length > body_buffer_size) {
    tmp_path = ...
    tmp_name.length = ...
    b = ...
} else {
    b = ...
}

if (body_length > body_buffer_size) {
    tmp_name.start = ...
    ...
    b->mem.start += sizeof(nxt_file_t) + tmp_name.length + 1;
    ...
}
```

I guess older GCC at least couldn't match those two if () statements and so couldn't tell that those two variables _had_ to be already initialised by the time we enter the second if () statement.

After your refactoring this is a complete non-issue anyway as we no longer make any reference to those two variables after the initital if () statement.

In fact those two variable should be declared in the if () statement

```diff
diff --git ./src/nxt_h1proto.c ./src/nxt_h1proto.c
index bb8022d5..40e865af 100644
--- ./src/nxt_h1proto.c
+++ ./src/nxt_h1proto.c
@@ -861,7 +861,6 @@ nxt_h1p_request_body_read(nxt_task_t *task, nxt_http_request_t *r)
 {
     size_t             size, body_length, body_buffer_size, body_rest;
     ssize_t            res;
-    nxt_str_t          *tmp_path, tmp_name;
     nxt_buf_t          *in, *b, *out, *chunk;
     nxt_conn_t         *c;
     nxt_h1proto_t      *h1p;
@@ -903,6 +902,8 @@ nxt_h1p_request_body_read(nxt_task_t *task, nxt_http_request_t *r)
                                body_length);
 
     if (r->chunked || (body_length > body_buffer_size)) {
+        nxt_str_t  *tmp_path, tmp_name;
+
         tmp_path = &r->conf->socket_conf->body_temp_path;
 
         tmp_name.length = tmp_path->length + tmp_name_pattern.length;
@@ -945,9 +946,6 @@ nxt_h1p_request_body_read(nxt_task_t *task, nxt_http_request_t *r)
         unlink((char *) tmp_name.start);
 
     } else {
-        tmp_path = NULL;
-        tmp_name.length = 0;
-
         b = nxt_buf_mem_alloc(r->mem_pool, body_buffer_size, 0);
         if (nxt_slow_path(b == NULL)) {
             status = NXT_HTTP_INTERNAL_SERVER_ERROR;
```

Reply:
Good to know, thanks.

> In fact those two variable should be declared in the if () statement

I'm fine with both the two ways, I'll keep it the same.

Reply:
I don't think leaving it as is makes sense...

Previously those dummy initialisations were to stop old versions of GCC from complaining [-Wuninitialized].

There was also a comment explaining this.

Your code re-jigs things, keeping the (now totally unnecessary) dummy initialisations but removing the comment explaining why they are there.

So you can see we are now in an undesirable situation.

The options in increasing order of correctness...

1) Put the comment back.

However the comment will need updating as it no longer makes sense...

a) It talks about a compiler version that simply cannot be used to build Unit now. You would need to check if it compiles on CentOS 7 with gcc 4.8.5 with those dummy initialisations removed without complaining or not to know whether the comment is even required.

b) However (a) above is a completely pointless exercise as it won't complain now in any case because of the next point.

2) Simply remove those dummy initialisations. They serve absolutely _no_ purpose. Those two variables are _only_ used in that first if () branch.
 
3) Do (2) but also reduce the scope of those two variables to that if () branch as I showed in the above patch.

So really the options are (2) or (3) with (3) being the obviously _correct_ one.

Reply:
Makes sense. Updated with the way (3).

Reply:
That's good, though I'm not sure why you did it in the last commit ("http: Support chunked request bodies")  and not this one ("http: Refactored nxt_h1p_request_body_read()")?

Reply:
Good catch, I did it in the wrong commit.
I have updated it, thanks.

Reply:
Oh yeah, that patch looks much better now! :)

You can add my

```
Reviewed-by: Andrew Clayton <a.clayton@nginx.com>
```

to that as well.